pkg/service/model.go

```diff
@@ -41,50 +112,273 @@ func (s *Service) ConvertToMDModel(ctx context.Context, fileUID uuid.UUID, calle
 	// Get the appropriate prefix for the file type
 	prefix := getFileTypePrefix(fileType)
 
+	// Prepare files to process (as a slice of base64 strings)
+	var filesToProcess []string
+	var err error
+
+	// For PDFs, split into batches; for other file types, process as a single item
+	if fileType == artifactpb.FileType_FILE_TYPE_PDF {
+		pdfData, err := base64.StdEncoding.DecodeString(fileBase64)
+		if err != nil {
+			return "", fmt.Errorf("decoding base64 PDF: %w", err)
+		}
+		filesToProcess, err = splitPDFIntoBatches(pdfData, 8)
+		if err != nil {
+			return "", fmt.Errorf("splitting PDF into batches: %w", err)
+		}
+	} else {
+		filesToProcess = []string{fileBase64}
+	}
+
+	// Process all files (or batches for PDFs)
+	results := make([]string, len(filesToProcess))
+	errCh := make(chan error, len(filesToProcess))
+
+	// Add a done channel to handle context cancellation
+	done := make(chan struct{})
+	defer close(done)
+
 	namespaceID := "admin"
-	req := &modelpb.TriggerNamespaceModelRequest{
-		NamespaceId: namespaceID,
-		ModelId:     ConvertDocToMDModelID,
-		Version:     ConvertDocToMDModelVersion,
-		TaskInputs: []*structpb.Struct{
-			{
-				Fields: map[string]*structpb.Value{
-					"data": {Kind: &structpb.Value_StructValue{
-						StructValue: &structpb.Struct{
+
+	// Start a goroutine to watch for context cancellation
+	go func() {
+		select {
+		case <-ctx.Done():
+			// If context is cancelled, put the error in the error channel
+			errCh <- ctx.Err()
+		case <-done:
+			// Operation completed normally
+			return
+		}
+	}()
+
+	// Create a worker pool with limited concurrency
+	type job struct {
+		idx int
+		b64 string
+	}
+
+	// Create a jobs channel with enough buffer to hold all jobs
+	jobs := make(chan job, len(filesToProcess))
+
+	// Fill the jobs channel with work
+	for i, b64 := range filesToProcess {
+		jobs <- job{idx: i, b64: b64}
+	}
+	close(jobs)
+
+	// Use wait group to track when all workers are done
+	var wg sync.WaitGroup
+
+	// Start a limited number of worker goroutines
+	workerCount := maxConcurrentWorkers
+	if len(filesToProcess) < workerCount {
+		workerCount = len(filesToProcess)
+	}
+
+	logger.Info("Starting worker pool", zap.Int("numWorkers", workerCount), zap.Int("totalBatches", len(filesToProcess)))
+
+	for i := 0; i < workerCount; i++ {
```

Comment:
nit: Go 1.22 introduced ranging over integers
```suggestion
	for i := range workerCount {
```